src/lib/blkalloc/blk_allocator.h

```diff
@@ -184,155 +151,42 @@ class BlkAllocator {
 
     virtual BlkAllocStatus alloc_contiguous(BlkId& bid) = 0;
     virtual BlkAllocStatus alloc(blk_count_t nblks, blk_alloc_hints const& hints, BlkId& out_blkid) = 0;
+    virtual BlkAllocStatus alloc_on_disk(BlkId const& bid) = 0;
+
     virtual void free(BlkId const& id) = 0;
+    virtual void free_on_disk(BlkId const& bid) = 0;
+
     virtual blk_num_t available_blks() const = 0;
     virtual blk_num_t get_freeable_nblks() const = 0;
     virtual blk_num_t get_defrag_nblks() const = 0;
     virtual blk_num_t get_used_blks() const = 0;
     virtual bool is_blk_alloced(BlkId const& b, bool use_lock = false) const = 0;
-    virtual std::string to_string() const = 0;
+    virtual bool is_blk_alloced_on_disk(BlkId const& b, bool use_lock = false) const = 0;
 
-    virtual void cp_flush(CP* cp); // TODO: it needs to be a pure virtual function after bitmap blkallocator is derived
-                                   // from base BlkAllocator;
-
-    sisl::Bitset* get_disk_bm_mutable() {
-        set_disk_bm_dirty();
-        return m_disk_bm.get();
-    }
-    const sisl::Bitset* get_disk_bm_const() const { return m_disk_bm.get(); };
-    sisl::Bitset* get_realtime_bm() { return m_realtime_bm.get(); }
-    const sisl::Bitset* get_realtime_bm() const { return m_realtime_bm.get(); }
-
-    bool need_flush_dirty_bm() const { return is_disk_bm_dirty; }
-
-    void set_disk_bm(std::unique_ptr< sisl::Bitset > recovered_bm);
-    BlkAllocPortion& get_blk_portion(blk_num_t portion_num) {
-        HS_DBG_ASSERT_LT(portion_num, get_num_portions(), "Portion num is not in range");
-        return m_blk_portions[portion_num];
-    }
-
-    virtual void inited();
+    virtual std::string to_string() const = 0;
+    virtual void cp_flush(CP* cp) = 0;
 
     void incr_alloced_blk_count(blk_count_t nblks) { m_alloced_blk_count.fetch_add(nblks, std::memory_order_relaxed); }
```

Comment:
since appendblkallocator can decude this m_alloced_blk_count and doesn't call incr/decr in its implementation, should we move these three apis (inc/dec/get) into bitmap_blk_allocator.h? It will be inaccurate anyway for append if we keep it here.

Reply:
Make sense, however, we may need to look out for future blk allocators which need this, but I agree for now I can move that there.